### PR TITLE
FIX: Resource cleanup on Python shutdown to avoid segfaults

### PR DESCRIPTION
**Pull request by bewithgaurav**
### Work Item / Issue Reference  

> [AB#38824](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/38824)

> GitHub Issue: #254 

-------------------------------------------------------------------
### Summary   
  
This pull request addresses a critical segfault issue that occurred during Python interpreter shutdown when using the `mssql_python` package, especially in scenarios with unclosed cursors, syntax errors, or concurrent operations. The changes introduce robust detection of Python's shutdown state to prevent unsafe resource cleanup and improve test coverage for these edge cases.

Key changes include:

**Python Shutdown Safety & Resource Cleanup**
* Added a centralized helper function `is_python_finalizing()` in `ddbc_bindings.cpp` to reliably detect when Python is shutting down or finalizing, using both `Py_IsInitialized()` and the `sys._is_finalizing` attribute for compatibility across Python versions.
* Updated the `LOG` function to skip logging if Python is shutting down, and to handle exceptions gracefully to avoid crashes during interpreter cleanup. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R665-R700) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R717-R725)
* Modified the `SqlHandle::free()` method to prevent freeing `SQL_HANDLE_STMT` handles during shutdown, as their parent DBC handles may already be freed, avoiding double-free and segfaults. Handles are marked as freed instead.

**Test Coverage for Shutdown and Concurrency Edge Cases**
* Added multiple new tests in `test_005_connection_cursor_lifecycle.py` to ensure no segfaults occur during Python shutdown, including:
  - Tests for syntax errors without explicit resource cleanup.
  - Tests for multiple concurrent cursor operations and abrupt interpreter exits.
  - Tests for closing connections with pending results and aggressive threading scenarios.

These changes significantly improve the stability of the `mssql_python` package in complex and error-prone scenarios, especially during interpreter shutdown and multithreaded usage.

### CODE REVIEW

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 1]
```diff
@@ -665,21 +665,43 @@ void HandleZeroColumnSizeAtFetch(SQLULEN& columnSize) {
...
+        }
+        
+        py::gil_scoped_acquire gil;  // <---- this ensures safe Python API usage
+
+        // Check if sys module is available and not finalizing
+        py::object sys_module = py::module_::import("sys");
+        if (!sys_module.is_none()) {
+            py::object finalizing_func = sys_module.attr("_is_finalizing");
+            if (!finalizing_func.is_none() && finalizing_func().cast<bool>()) {
+                return; // Python is finalizing, don't log
```
[the remark this answers is not in the document]
Comment (bewithgaurav) [reply mssql_python/pybind/ddbc_bindings.cpp]: we support python 10+, not a valid comment

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 2]
```diff
@@ -662,24 +662,66 @@ void HandleZeroColumnSizeAtFetch(SQLULEN& columnSize) {
...
+            // Check if the attribute exists before accessing it (for Python version compatibility)
+            if (py::hasattr(sys_module, "_is_finalizing")) {
+                py::object finalizing_func = sys_module.attr("_is_finalizing");
+                if (!finalizing_func.is_none() && finalizing_func().cast<bool>()) {
+                    return true; // Python is finalizing
+                }
+            }
+        }
+        return false;
+    } catch (...) {
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:688]: Can we consider logging or at least debugging such exceptions if they become frequent (possibly with a debug flag)?

Comment (bewithgaurav) [reply mssql_python/pybind/ddbc_bindings.cpp:688]: Actually, its not recommended to put any loggers on python shutdown, so we can discard putting our standard loggers here.
This catch is in case python doesn't get what is `sys` module etc., so this is absolutely unexpected but viable during shutdown scenarios. This catch further prevents a core dump. 
But this raises a good point, we or the user does need to know if we're getting unexpected exceptions during shutdown.
I guess we can put an std::cerr here which will do an stdout instead of logging, will make a small change here.

Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp:688]: Yes, that makes sense. Using std::cerr here is a good compromise, it avoids unsafe logging during Python shutdown while still surfacing unexpected exceptions for debugging.

**Review by sumitmsft — APPROVED** [review]
LGTM. Approved
